### ISSUE TITLE
[BUG] SIGSEGV in notifyDataAvailable during `curl_easy_pause`

### ISSUE THREAD

**Issue by schultetwin1** [awslabs/amazon-kinesis-video-streams-producer-c#118]
**Describe the bug**
We are seeing a sporadic SIGSEGV (write to address 0xC, so probably a NULL de-reference) during a call to putKinesisVideoFrame. Here is the full stack trace of the faulting thread (note, only exported symbols available so the internal calls in libcurl are not shown).

```
1.  0x772bbbb0 at malloc + 704                         libc.so.0
2.  0x773e4224 at ???                                  libcurl.so.4
3.  0x773e5b84 at ???                                  libcurl.so.4
4.  0x773dec74 at curl_easy_pause + 280                libcurl.so.4
5.  0x7735b07c at notifyDataAvailable + 256            libcproducer.so
6.  0x77353a20 at dataAvailableCurl + 224              libcproducer.so
7.  0x773504b4 at streamDataAvailableAggregate + 180   libcproducer.so
8.  0x7733625c at putFrame + 3168                      libcproducer.so
9.  0x7733b3a8 at putEarliestFrame + 260               libcproducer.so
10. 0x7733b784 at frameOrderCoordinatorPutFrame + 676  libcproducer.so
11. 0x7732d988 at putKinesisVideoFrame + 80            libcproducer.so
```

And here is a stack trace of one of the helper threads (waiting to get the glocal malloc lock)

```
1. 0x7737dcf0 at ???                                   libpthread.so
2. 0x77385068 at pthread_mutex_lock + 280              libpthread.so
3. 0x772bcbcc at free + 76                             libc.so
4. 0x773e5218 at curl_multi_wait + 724                 libcurl.so
5. 0x773de5f8 at curl_easy_perform + 336               libcurl.so
6. 0x7735ad1c at curlCompleteSync + 452                libcproducer.so
7. 0x773545dc at putStreamCurlHandler + 320            libcproducer.so
```

Based on the code in `notifyDataAvailable` it looks like this is a known issue which has been worked around using a `sleep` call. 

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/7e0d30f83e41cc9c565bd45d277be6a2c1bb86f7/src/source/Response.c#L483-L489

Our core dumps currently do not have symbols, but based on some initial investigations, it looks like a double free is occurring somewhere.

We are using version 2.0.1.

I know that's not much to go on, and I will try and get more information. Let me know if something else might help.

**Comment by MushMal** [comment]
The first frame in the crash stack is the malloc. Where did you get the crash info on 0xc? If the reference registers point to 0x0c then it is a null deref but the top of the stack being malloc makes me wonder if we have a memory corruption issue somewhere.

V 2.0.1 is fairly old at this stage. I wonder if you are able to reproduce this on later commits. Also, a debug crash stack or a stack with symbols for all of the threads might be helpful in conjunction with the debug logs to see what's going on.

**Comment by staowyze** [comment]
is the [curl segfault comment](https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/master/src/source/Response.c#L483) still valid in latest master? it sounds scary if we try to sleep a bit to avoid crash.

**Comment by MushMal** [comment]
To be honest, I am not sure what exact condition is hit with the offline case to prompt this sleep. Perhaps, it's benign at this stage. At least, none of the unit tests are hitting this. 

Just like with libWebSockets, the shutdown sequence for curl is left to the application to handle. The complexity comes from the fact that the networking clients have to deal with OS-level resources that need to be properly freed on the same thread. So, an application calling a shutdown (from it's own main thread) can't directly "shutdown" the ongoing ops.

**Comment by MushMal** [comment]
Any further info on this?

**Comment by schultetwin1** [comment]
Sorry for the delay, unfortunately we don't have more information on this. One question for you:

> Perhaps, it's [the thread sleep] benign at this stage. At least, none of the unit tests are hitting this.

Since we don't understand why the sleep is there in the first place, would you feel confident in removing this sleep?

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/7e0d30f83e41cc9c565bd45d277be6a2c1bb86f7/src/source/Response.c#L483-L489

This would either tell us if this sleep is actually benign or maybe make this issue more reproducible

**Comment by MushMal** [comment]
Try to remove it in your local build and run your scenario. Please try to capture the crash stack if you hit it. The offline case is not very specific to this - just the case of high frequency pause I pause that could cause issues with the libCurl internal event loop.

**Comment by schultetwin1** [comment]
> Try to remove it in your local build and run your scenario.

That's a good idea, we will give it a try. @MushMal, are you an AWS employee or just helping out in your spare time? Your GitHub profile is not a part of the Amazon Web Services - Labs organization.

**Comment by MushMal** [comment]
That's odd - it should be listed.

**Comment by MushMal** [comment]
Closing this issue. Please feel free to reactivate or cut a new one if you have any issues

**Comment by schultetwin1** [comment]
@MushMal - We were able to capture a better stack trace. It looks like there is a race condition between two threads both operating on the same `CURL*`

Here is the stack of the crashing thread (let's call it Thread 1)

```
libcurl.so.4!Curl_llist_insert_next(struct Curl_llist * list, struct Curl_llist_element * e, const void * p, struct Curl_llist_element * ne) (/home/matt/dev/curl/lib/llist.c:73)
libcurl.so.4!multi_addtimeout(expire_id eid, struct curltime * stamp, struct Curl_easy * data) (/home/matt/dev/curl/lib/multi.c:3173)
libcurl.so.4!Curl_expire(struct Curl_easy * data, timediff_t milli, expire_id id) (/home/matt/dev/curl/lib/multi.c:3215)
libcurl.so.4!curl_easy_pause(struct Curl_easy * data, int action) (/home/matt/dev/curl/lib/easy.c:1081)
libcproducer.so!notifyDataAvailable(PCurlResponse pCurlResponse, UINT64 durationAvailable, UINT64 sizeAvailable) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/src/source/Response.c:462)
libcproducer.so!dataAvailableCurl(UINT64 customData, STREAM_HANDLE streamHandle, PCHAR streamName, UPLOAD_HANDLE uploadHandle, UINT64 durationAvailable, UINT64 bytesAvailable) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/src/source/CurlApiCallbacks.c:815)
libcproducer.so!streamDataAvailableAggregate(UINT64 customData, STREAM_HANDLE streamHandle, PCHAR streamName, UPLOAD_HANDLE uploadHandle, UINT64 availableDuration, UINT64 availableSize) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/src/source/CallbacksProvider.c:1275)
libcproducer.so!putFrame(PKinesisVideoStream pKinesisVideoStream, PFrame pFrame) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/dependency/libkvspic/kvspic-src/src/client/src/Stream.c:1036)
libcproducer.so!putEarliestFrame(PKinesisVideoStream pKinesisVideoStream) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/dependency/libkvspic/kvspic-src/src/client/src/FrameOrderCoordinator.c:235)
libcproducer.so!frameOrderCoordinatorPutFrame(PKinesisVideoStream pKinesisVideoStream, PFrame pUserFrame) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/dependency/libkvspic/kvspic-src/src/client/src/FrameOrderCoordinator.c:321)
libcproducer.so!putKinesisVideoFrame(STREAM_HANDLE streamHandle, PFrame pFrame) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/dependency/libkvspic/kvspic-src/src/client/src/Client.c:781)
put_stream_to_kinesis_handler(frame_head_s * head) (Unknown Source:0
receive_stream_from_shm() (Unknown Source:0)
receive_stream_function(void * arg) (Unknown Source:0)
libpthread.so.0!pthread_cond_destroy@@GLIBC_2.3.2 (Unknown Source:0)
[Unknown/Just-In-Time compiled code] (Unknown Source:0)
```

And here is the stack of the second thread operating on the same `CURL*` object (let's call it Thread 2)

```
libpthread.so.0!__lll_lock_wait (Unknown Source:0)
libpthread.so.0!pthread_mutex_lock (Unknown Source:0)
libcproducer.so!getStreamData(PKinesisVideoStream pKinesisVideoStream, UPLOAD_HANDLE uploadHandle, PBYTE pBuffer, UINT32 bufferSize, PUINT32 pFillSize) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/dependency/libkvspic/kvspic-src/src/client/src/Stream.c:1127)
libcproducer.so!getKinesisVideoStreamData(STREAM_HANDLE streamHandle, UPLOAD_HANDLE uploadHandle, PBYTE pBuffer, UINT32 bufferSize, PUINT32 pFilledSize) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/dependency/libkvspic/kvspic-src/src/client/src/Client.c:904)
libcproducer.so!postReadCallback(PCHAR pBuffer, SIZE_T size, SIZE_T numItems, PVOID customData) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/src/source/Response.c:638)
libcurl.so.4!Curl_fillreadbuffer(struct connectdata * conn, size_t bytes, size_t * nreadp) (/home/matt/dev/curl/lib/transfer.c:245)
libcurl.so.4!readwrite_upload(int * didwhat, struct connectdata * conn, struct Curl_easy * data) (/home/matt/dev/curl/lib/transfer.c:1057)
libcurl.so.4!Curl_readwrite(struct connectdata * conn, struct Curl_easy * data, _Bool * done, _Bool * comeback) (/home/matt/dev/curl/lib/transfer.c:1263)
libcurl.so.4!multi_runsingle(struct Curl_multi * multi, struct curltime * nowp, struct Curl_easy * data) (/home/matt/dev/curl/lib/multi.c:2147)
libcurl.so.4!curl_multi_perform(struct Curl_multi * multi, int * running_handles) (/home/matt/dev/curl/lib/multi.c:2417)
libcurl.so.4!easy_transfer() (/home/matt/dev/curl/lib/easy.c:593)
libcurl.so.4!easy_perform(struct Curl_easy * data) (/home/matt/dev/curl/lib/easy.c:683)
libcurl.so.4!curl_easy_perform(struct Curl_easy * data) (/home/matt/dev/curl/lib/easy.c:702)
libcproducer.so!curlCompleteSync(PCurlResponse pCurlResponse) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/src/source/Response.c:405)
libcproducer.so!curlCallApi(PCurlRequest pCurlRequest) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/src/source/CurlApiCallbacks.c:650)
libcproducer.so!putStreamCurlHandler(PVOID arg) (/home/matt/dev/rootfs/build/producer-c/src/producer-c/src/source/CurlApiCallbacks.c:2136)
libpthread.so.0!start_thread (Unknown Source:0)
libpthread.so.0!__thread_start (Unknown Source:0)
```

Notice that Thread 1 (the crashing thread) is in `curl_easy_pause` while Thread 2 is in `curl_easy_perform`. The pointers passed into both `curl_easy_pause` and `curl_easy_perform` is the same. According to [libcurl's website](https://curl.se/libcurl/c/threadsafe.html), handles are not thread safe:

>  You must never share the same handle in multiple threads. You can pass the handles around among threads, but you must never use a single handle from more than one thread at any given time.

Since there two threads are both operating on the same CURL handle, I'm guessing that's what is causing the segfault.

If it helps, the `timerQueueExecutor` thread, is currently paused waiting for the next time to execute.

I'm having a hard time working through the state machines to determine why one thread would be "un-pausing" the CURL handle while the other thread is using it to upload data.

We cannot repro this issue, but we are collecting core dumps which is where we found this race.

**Comment by MushMal** [comment]
@schultetwin1 I am not sure yet what's going on in your case. Is your application streaming two or more separate streams? If so, then those are different curl objects. 

The pausing/unpausing is gated by response->paused boolean. This should really become an atomic variable. @hassanctech the definition of paused located here

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/master/src/source/Response.h#L65

has a volatile type to avoid variable shearing but I don't see how that could be a factor here.  
Given the stack trace is for an older codebase it's hard to guess what exactly is going on. Perhaps it has been fixed in the latest?

@schultetwin1 how often is this hit? Is there a way to beta the latest and see if you hit this in your app?

The second thread stack is attempting to acquire the stream lock. I also can't think of a good reason why you have this stack unless you are looking at two separate streams.

**Comment by schultetwin1** [comment]
> Is your application streaming two or more separate streams? If so, then those are different curl objects.

We are only streaming one stream (with two tracks). In both threads, `pCurlRequest` points to the same location (`0x542988`) and `pCurlRequest->pCurlResponse->pCurl` also both point to the same location (`0x53050c`), which is why I think they are the same cURL handle.

> The pausing/unpausing is gated by response->paused boolean. This should really become an atomic variable.

There is also a place in the code where `paused` is set to `FALSE` but then the actual unpausing does not occur for 10ms. Could there be a race between here and the consumer of the `paused` flag?

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/a8091d1a140fbdcb260ba3c8302aa2e4712d2d91/src/source/Response.c#L457-L466

> how often is this hit?

That's a good question. We are working on getting those numbers but currently have no automation running on collected core dumps.

> Is there a way to beta the latest and see if you hit this in your app?

This stack trace is from your latest released version (1.1.1). We, unfortunately, have no way to reproduce this issue and so would need to release new firmware and wait for more core dumps to test if the latest has fixed this issue. We'd only want to release firmware which has an officially released by AWS, based on the [guidance found in the README.md](https://github.com/awslabs/amazon-kinesis-video-streams-producer-c#release). Is a new release scheduled for anytime soon?

**Comment by MushMal** [comment]
Thanks for the clarification. I am somewhat puzzled. Under normal circumstances (single stream) there should be really two threads operating - one that's producing the frame into the SDK, the media thread which calls the data available notification callback and the other one which is driving the networking. In your stacks, the first one is the thread stack is the media thread that produces the frame which in turn drives the data notification callback. It checks for the paused flag and tries to unpause if paused so the curl could enable the reader callback on the network thread which is what's on your second stack. Looking at the current master code I am having hard time understanding how this could happen. 

Looking at the changes after the 1.1.1 release I could see this commit which shouldn't really impact this issue but is should have a large impact timing-wise

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/commit/a8091d1a140fbdcb260ba3c8302aa2e4712d2d91

Perhaps @hassanctech can have a better answer.

**Comment by hassanctech** [comment]
I don't see a race condition here but I'm thinking about something...

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/f2a97fe6eaf78cbffd46ccfa5994bee2bebf99bf/src/source/Response.c#L34

We initialize paused to `TRUE` this is more prescriptive of what we want to do than descriptive of the current state. The result here is basically when curl invokes the post read callback, in the callback we check for this: 

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/f2a97fe6eaf78cbffd46ccfa5994bee2bebf99bf/src/source/Response.c#L628

This will now result in the curl being paused.  Under normal operation after data is flowing, i.e. after the `notifyDataAvailable` has fired with curl paused and it has toggled the value of `paused` to `FALSE`, now we will never enter the post read callback in a case where `paused` is `TRUE.

But based on the stack trace what's happening is when the post read callback enters the value of `paused` must already be FALSE or else the function would have already returned (based on the block starting at line 34 I posted above), so then it enters the getKinesisVideoStreamData call which ultimately is waiting on the streams lock (Thread 2).  In the mean time Thread 1 waits for 10ms (after it sets `paused` to FALSE then proceeds to unpause). Perhaps curl unpause crashes if it's not actually paused?  I'm not 100% sure but maybe we're banking on the post read callback to be invoked by libcurl *faster* than us calling `notifyDataAvailable` to avoid this situation and maybe that's why that 10ms pause was there to make sure the post read callback was called at least once before us calling unpause to make sure it had a returned `CURL_READFUNC_PAUSE` and already paused curl.

In any case I need to think about this a little bit more but I'm thinking we should initialize the value of `paused` to `FALSE`.  In the latest code (head of master) the check in the post read callback has actually been removed already.

There's another way this can happen after streaming has commenced, which is in the post read callback we set `paused` to `TRUE` here:  

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/f2a97fe6eaf78cbffd46ccfa5994bee2bebf99bf/src/source/Response.c#L707

Now suppose we're at that line and at the same time `notifyDataAvailable` is invoked and we get to here:

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/f2a97fe6eaf78cbffd46ccfa5994bee2bebf99bf/src/source/Response.c#L457

It's `TRUE` so we enter the conditional and make it `FALSE` then we call unpause after 10ms.  But that 10ms gives us the grace period we need for the post read callback to have actually returned `CURL_READFUNC_PAUSE` and for curl to have gone into pause state so nothing bad happens here.  BUT now `paused` is FALSE but curl IS actually paused.  In this case we will never invoke unpause but that doesn't look like what's happening.  

We can't make setting the boolean and actually putting curl into pause/unpause an an atomic operation to prevent them from getting out of sync.  This is because in the case of returning `CURL_READFUNC_PAUSE` from the post read callback the pausing is done internally so we can't make that part atomic, or even put a lock around it.  We could for un-pause since we call that explicitly but it doesn't help unless we can do both.  

There's a TODO/nice to have open for curl to implement a feature to be able to return the pause state of a transfer but unfortunately it doesn't exist now:  https://github.com/curl/curl/issues/2588.  I'll have to give some more thought how to avoid this.

Two asks, @schultetwin1. The frequency with which curl pause is called should be decreased drastically (under decent network conditions) with the latest commit (head of master) would it be possible to try that?  Unless your network is taking > 600ms for any data to be available / putFrame to be called, you should pretty much never be pausing transfers so significantly decreasing the likelihood of hitting this.  Which version of libcurl are you using?

**Comment by MushMal** [comment]
Yeah, unfortunately the easy interface is very limited and it's terrible that we can't get the paused state from the source of truth and have to cache it and get out-of-sync. I am really suspicious about the 10ms pause - this just doesn't seem right. I am sure this was added a while back empirically looking at the crash stacks but I doubt it's needed. Perhaps, removing it and putting it through a stress might shed more light. It's easy to stress it - have a high frame rate of very low bitrate stream.

**Comment by schultetwin1** [comment]
Thank you both @MushMal and @hassanctech for your detailed analysis! First, to answer your questions:

> The frequency with which curl pause is called should be decreased drastically (under decent network conditions) with the latest commit (head of master) would it be possible to try that

We, unfortunately, are not able to reproduce this crash locally. We found this crash from inspecting crashes occurring on devices running our beta firmware. I'd prefer to wait for another official release of the producer-c SDK before deploying an updated version of the producer-c SDK to our beta devices to reduce the risk of introducing any new issue.

> Which version of libcurl are you using?

We are using 7.74.0.

> Unless your network is taking > 600ms for any data to be available / putFrame to be called

We are using a intermittent producer, so it could be possible that there is a >600ms delay between to calls to `putFrame`.
